### DISCUSSION TITLE
How to replace torch.nn.CrossEntropyLoss by monai.losses.FocalLoss?

### DISCUSSION THREAD

**Discussion by dzenanz** [Project-MONAI/MONAI#2027] [General]
I am trying to check whether focal loss improves the training of my network. Just replacing CrossEntropyLoss by FocalLoss leads to an error.

When I set `use_focal_loss = True`
```py
if use_focal_loss:
    loss_function = monai.losses.FocalLoss(weight=class_weights)
else:
    loss_function = torch.nn.CrossEntropyLoss(weight=class_weights)
```
I get: `ValueError: logits and target ndim must match, got logits=2 target=1.`

If I also add at the end of `forward`:
```py
if use_focal_loss:
    outputs = torch.nn.functional.log_softmax(outputs, dim=1)
```
the error message is unchanged.

**Comment by dzenanz** [comment]
Anyone?

**Reply by wyli** [reply]
perhaps set `to_one_hot_y=True`?
example:
```py
>>> import torch
>>> from monai.losses import FocalLoss
>>> pred = torch.tensor([[1, 0], [0, 1], [1, 0]], dtype=torch.float32)
>>> grnd = torch.tensor([0, 1, 0], dtype=torch.int64)
>>> fl = FocalLoss(to_onehot_y=True)
>>> fl(pred, grnd)
tensor(0.0113)
>>> pred.shape
torch.Size([3, 2])
>>> grnd.shape
torch.Size([3])
>>> 
```

**Reply by dzenanz** [reply]
With MONAI 0.4:

    loss_function = monai.losses.FocalLoss(weight=class_weights, to_onehot_y=True)
TypeError: __init__() got an unexpected keyword argument 'to_onehot_y'

With MONAI 0.5 it works. Thanks a lot Wenqi!
